### PR TITLE
fix for token claiming

### PR DESCRIPTION
**Pull request by debjit-bhowal-rs**

### CODE REVIEW

File: .github/workflows/deploy-preview.yml [hunk 1]
```diff
@@ -52,7 +52,7 @@ jobs:
           flyctl secrets set GRPC_AUTH_TOKEN="$GRPC_AUTH_TOKEN" --app "$APP_NAME" --stage
           flyctl secrets set HOTORNOT_GOOGLE_CLIENT_SECRET="$HOTORNOT_GOOGLE_CLIENT_SECRET" --app "$APP_NAME" --stage
           flyctl secrets set ICPUMPFUN_GOOGLE_CLIENT_SECRET="$ICPUMPFUN_GOOGLE_CLIENT_SECRET" --app "$APP_NAME" --stage
-          flyctl deploy --app $APP_NAME
+          flyctl deploy --app $APP_NAME --depot=false
```
Comment (rupansh-gob) [inline .github/workflows/deploy-preview.yml:55]: This should not be necessary? cc: @ravibazz

File: ssr/src/component/claim_tokens.rs [hunk 2]
```diff
@@ -1,24 +1,89 @@
...
+                        <p>{format!("{} {}", info.balance.humanize(), info.symbol)}</p>
+                    }
+                })
+            }}
+        </Suspense>
+    }
+}
+
+#[component]
+pub fn ClaimTokensOrRedirectError(
```
Comment (rupansh-gob) [inline ssr/src/component/claim_tokens.rs:48]: Rename to ClaimTokenAndFetchBalance

File: ssr/src/component/claim_tokens.rs [hunk 3]
```diff
@@ -1,24 +1,89 @@
...
+            let res = claim_tokens_from_first_neuron_if_required(
+                cans_wire.clone(),
+                token_root,
+                user_principal,
+                true,
+            )
+            .await;
+            match res {
+                Ok(_) => Ok(()),  // No claim required, token balance can be displayed
+                Err(e) => Err(e), // Claim required or another error occurred
```
Comment (rupansh-gob) [inline ssr/src/component/claim_tokens.rs:69]: This misuses the result type

File: ssr/src/component/claim_tokens.rs [hunk 4]
```diff
@@ -1,24 +1,89 @@
...
+            }
+        },
+    );
+
     let claim_res = auth_cans.derive(
```
Comment (rupansh-gob) [inline ssr/src/component/claim_tokens.rs:74]: `claim_res` should be created only when required,
Resources execute as soon as you create them.

File: ssr/src/component/claim_tokens.rs [hunk 5]
```diff
@@ -33,15 +98,36 @@ pub fn ClaimTokensOrRedirectError(token_root: Principal) -> impl IntoView {
...
+                                                    view! {<TokenDisplay user_principal=user_principal token_root=token_root/>}
+                                                },
+                                                Err(e) => {
+                                                    view! {<Redirect path=format!("/error?err={e}")/>}
+                                                }
+                                            }
+                                        })
+                                    }}
+                                </Suspense>
+                            }
```
Comment (rupansh-gob) [inline ssr/src/component/claim_tokens.rs:126]: Move to a separate component?

File: ssr/src/page/profile/tokens.rs [hunk 6]
```diff
@@ -18,6 +18,24 @@ fn TokenViewFallback() -> impl IntoView {
...
+
+#[component]
+fn TokenClaiming(info: TokenMetadata) -> impl IntoView {
+    view! {
+        <span class="text-white truncate">
+        {format!("{} {}", "Claiming", info.symbol)}
+        </span>
+    }
+}
+
```
Comment (rupansh-gob) [inline ssr/src/page/profile/tokens.rs:38]: PLEASE make it a habit to not push dead code

File: ssr/src/page/wallet/tokens.rs [hunk 7]
```diff
@@ -56,7 +56,7 @@ impl CursoredDataProvider for TokenRootList {
     }
 }
 
-async fn token_metadata_or_fallback(
+pub async fn token_metadata_or_fallback(
```
Comment (rupansh-gob) [inline ssr/src/page/wallet/tokens.rs:59]: Move this to utils/token

File: ssr/src/component/claim_tokens.rs [hunk 8]
```diff
@@ -1,24 +1,89 @@
...
+#[component]
+fn TokenDisplay(user_principal: Principal, token_root: Principal) -> impl IntoView {
+    let cans = unauth_canisters();
+    let info = create_resource(
+        || (),
+        move |_| token_metadata_or_fallback(cans.clone(), user_principal, token_root),
+    );
+
+    view! {
+        <Suspense fallback=TokenClaimFallback>
```
Comment (rupansh-gob) [inline ssr/src/component/claim_tokens.rs:35]: I think we discussed we don't want to refetch balance when its not required?

File: ssr/src/utils/token/mod.rs [hunk 9]
```diff
@@ -281,6 +288,10 @@ pub async fn claim_tokens_from_first_neuron_if_required(
         return Ok(());
     }
 
+    if only_check {
+        // this prevents claiming during initial checking to prevent races
+        return Err(ServerFnError::ServerError("Claim required".to_string()));
+    }
```
Comment (rupansh-gob) [inline ssr/src/utils/token/mod.rs:294]: Create a separate function which returns `Result<Option<Neuron>>` that must be claimed, Then `claim_tokens_from_first_neuron` can be called directly in the new component

**Review by rupansh-gob — CHANGES_REQUESTED** [review]

File: .github/workflows/deploy-preview.yml (hunk 1 above)
Comment (debjit-bhowal-rs) [reply .github/workflows/deploy-preview.yml:55]: yep, was temp, will undo

File: ssr/src/component/claim_tokens.rs (hunk 5 above)
Comment (debjit-bhowal-rs) [reply ssr/src/component/claim_tokens.rs:126]: done

**Review by rupansh-sekar-yral — APPROVED** [review]

**Pull request merged**